examples/grad3d/grad3d.f90

```diff
@@ -0,0 +1,625 @@
+
+      implicit none
+
+      ! Arguments
+      integer, intent(in) :: nx, ny, nz
+      real(mytype), intent(in) :: delta
+      real(mytype), intent(out), dimension(nx,ny,nz) :: df
+      real(mytype), intent(in), dimension(nx,ny,nz) :: ff
+
+      ! Local variables
+      integer :: i, j, k
+      real(mytype) :: coeff = 0.5_mytype
+
+      coeff = coeff/delta
+
+      !$acc kernels default(present)
+      do concurrent (k=1:nz, j=1:ny)
+         df(1,j,k) = coeff*(ff(2,j,k)-ff(nx,j,k)) 
+         do concurrent (i=2:nx-1)
+            df(i,j,k) = coeff*(ff(i+1,j,k)-ff(i-1,j,k)) 
+         enddo
+         df(nx,j,k) = coeff*(ff(1,j,k)-ff(nx-1,j,k))
+      enddo
+      !$acc end kernels
+
+   end subroutine derx 
+   !=====================================================================
+   ! Calculate gradient in Y (data in Y-pencil)
+   !=====================================================================
+   subroutine dery(df,ff,delta,nx,ny,nz)
+
+      implicit none
+
+      ! Arguments
+      integer, intent(in) :: nx, ny, nz
+      real(mytype), intent(in) :: delta
+      real(mytype), intent(out), dimension(nx,ny,nz) :: df
+      real(mytype), intent(in), dimension(nx,ny,nz) :: ff
+
+      ! Local variables
+      integer :: i, j, k
+      real(mytype) :: coeff = 0.5_mytype
+
+      coeff = coeff/delta
+
+      !$acc kernels default(present)
+      do concurrent (k=1:nz)
+        do concurrent (i=1:nx)
+           df(i,1,k) = coeff*(ff(i,2,k)-ff(i,ny,k))
+        enddo
+        do concurrent (j=2:ny-1, i=1:nx)
+           df(i,j,k) = coeff*(ff(i,j+1,k)-ff(i,j-1,k))
+        enddo
+        do concurrent (i=1:nx)
+           df(i,ny,k) = coeff*(ff(i,1,k)-ff(i,ny-1,k)) 
+        enddo
+      enddo
+      !$acc end kernels
+
+   end subroutine dery 
+   !=====================================================================
+   ! Calculate gradient in Z (data in Z-pencil)
+   !=====================================================================
+   subroutine derz(df,ff,delta,nx,ny,nz)
+
+      implicit none
+
+      ! Arguments
+      integer, intent(in) :: nx, ny, nz
+      real(mytype), intent(in) :: delta
+      real(mytype), intent(out), dimension(nx,ny,nz) :: df
+      real(mytype), intent(in), dimension(nx,ny,nz) :: ff
+
+      ! Local variables
+      integer :: i, j, k
+      real(mytype) :: coeff = 0.5_mytype
+      
+      coeff = coeff/delta
+
+      !$acc kernels default(present)
+      do concurrent (j=1:ny, i=1:nx)
+         df(i,j,1) = coeff*(ff(i,j,2)-ff(i,j,nz  ))
+      enddo
+      do concurrent (k=2:nz-1, j=1:ny, i=1:nx)
+         df(i,j,k) = coeff*(ff(i,j,k+1)-ff(i,j,k-1))
+      enddo
+      do concurrent (j=1:ny, i=1:nx)
+         df(i,j,nz) = coeff*(ff(i,j,1)-ff(i,j,nz-1))
+      enddo
+      !$acc end kernels
+
+   end subroutine derz 
+   !=====================================================================
+   ! Calculate divergence using halo-cell exchange (data in X-pencil)
```

Comment:
Test derivatives against analytical solution (data in X-pencil)

Reply:
Done